File: frontend/src/types/asset.ts
// asset.ts — 资产管理模块前端类型定义

// ── 通用响应包装 ──

export interface Result<T> {
  success: boolean
  data: T
  message: string
}

// ── 环境 / 分组 / 凭据 ──

export interface Environment {
  id: number
  name: string
  description: string
  color: string
  created_at: string
  updated_at: string
  groups?: Group[]
}

export interface Group {
  id: number
  environment_id: number
  name: string
  description: string
  created_at: string
  updated_at: string
  environment?: Environment
}

export type CredentialType = 'password' | 'ssh_key' | 'token'

export interface Credential {
  id: number
  name: string
  type: CredentialType
  username: string
  secret_masked?: string
  created_at: string
  updated_at: string
}

// ── 插件 Schema ──

export type ConfigFieldType =
  | 'text'
  | 'number'
  | 'password'
  | 'textarea'
  | 'boolean'
  | 'select'

export interface SelectOption {
  value: string
  label: string
}

export interface ConfigField {
  key: string
  label: string
  type: ConfigFieldType
  required: boolean
  default_val?: unknown
  options?: SelectOption[]
  placeholder?: string
  description?: string
  secret?: boolean
}

export type AssetCategory = 'server' | 'database' | 'cache' | 'mq' | 'other'

export interface PluginDef {
  type_id: string
  display_name: string
  category: AssetCategory
  icon_name: string
  config_schema: ConfigField[]
}

// ── 资产 ──

export type AssetStatus = 'unknown' | 'online' | 'offline' | 'warning'

export interface Asset {
  id: number
  environment_id: number
  group_id?: number
  category: AssetCategory
  plugin_type: string
  name: string
  description: string
  tags: string[]
  credential_id?: number
  status: AssetStatus
  last_checked_at?: string
  ext_config: Record<string, unknown>
  created_at: string
  updated_at: string
  environment?: Environment
  group?: Group
  credential?: Credential
}

// ── 请求体类型 ──

export interface CreateEnvironmentReq {
  name: string
  description: string
  color: string
}

export interface UpdateEnvironmentReq extends CreateEnvironmentReq {
  id: number
}

export interface CreateGroupReq {
  environment_id: number
  name: string
  description: string
}

export interface UpdateGroupReq {
  id: number
  name: string
  description: string
}

export interface CreateAssetReq {
  environment_id: number
  group_id?: number
  category: AssetCategory
  plugin_type: string
  name: string
  description: string
  tags: string[]
  credential_id?: number
  ext_config: Record<string, unknown>
}

export interface UpdateAssetReq {
  id: number
  group_id?: number
  name: string
  description: string
  tags: string[]
  credential_id?: number
  ext_config: Record<string, unknown>
}

export interface ListAssetsReq {
  environment_id?: number
  group_id?: number
  category?: AssetCategory | ''
  plugin_type?: string
  keyword?: string
}

export interface CreateCredentialReq {
  name: string
  type: CredentialType
  username: string
  secret: string
}

export interface UpdateCredentialReq {
  id: number
  name: string
  type: CredentialType
  username: string
  secret: string
}

// ── 展示辅助常量 ──

export const CATEGORY_LABELS: Record<AssetCategory, string> = {
  server: '服务器',
  database: '数据库',
  cache: '缓存',
  mq: '消息队列',
  other: '其他',
}

export const CATEGORY_COLORS: Record<AssetCategory, string> = {
  server: '#60a5fa',
  database: '#f97316',
  cache: '#ef4444',
  mq: '#8b5cf6',
  other: '#6b7280',
}

export const ASSET_STATUS_LABELS: Record<AssetStatus, string> = {
  unknown: '未检测',
  online: '在线',
  offline: '离线',
  warning: '告警',
}

export const ASSET_STATUS_COLORS: Record<AssetStatus, string> = {
  unknown: '#6b7280',
  online: '#22c55e',
  offline: '#ef4444',
  warning: '#f59e0b',
}

/** 从 ext_config 中提取可展示的连接地址 */
export function getAssetAddress(asset: Asset): string {
  const cfg = asset.ext_config ?? {}
  const host = cfg['host'] as string | undefined
  const port = cfg['port'] as number | undefined
  if (host && port) return `${host}:${port}`
  if (host) return host
  // RocketMQ
  const ns = cfg['name_server'] as string | undefined
  if (ns) return ns
  // Kafka
  const brokers = cfg['brokers'] as string | undefined
  if (brokers) return brokers
  return '—'
}

File: internal/asset/model/asset.go
package model

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"time"

	"EnvPilot/internal/plugin"
)

// AssetStatus 资产连接状态
type AssetStatus string

const (
	AssetStatusUnknown AssetStatus = "unknown"
	AssetStatusOnline  AssetStatus = "online"
	AssetStatusOffline AssetStatus = "offline"
	AssetStatusWarning AssetStatus = "warning"
)

// Tags 标签列表，存储为 JSON 字符串
type Tags []string

func (t Tags) Value() (driver.Value, error) {
	if len(t) == 0 {
		return "[]", nil
	}
	b, err := json.Marshal(t)
	return string(b), err
}

func (t *Tags) Scan(value interface{}) error {
	var data []byte
	switch v := value.(type) {
	case string:
		data = []byte(v)
	case []byte:
		data = v
	case nil:
		*t = Tags{}
		return nil
	default:
		return fmt.Errorf("Tags.Scan: 不支持的类型 %T", value)
	}
	return json.Unmarshal(data, t)
}

// ExtConfig 类型专属扩展配置，以 JSON 存储在单列中
type ExtConfig map[string]interface{}

func (e ExtConfig) Value() (driver.Value, error) {
	if e == nil {
		return "{}", nil
	}
	b, err := json.Marshal(e)
	return string(b), err
}

func (e *ExtConfig) Scan(value interface{}) error {
	var data []byte
	switch v := value.(type) {
	case string:
		data = []byte(v)
	case []byte:
		data = v
	case nil:
		*e = make(ExtConfig)
		return nil
	default:
		return fmt.Errorf("ExtConfig.Scan: 不支持的类型 %T", value)
	}
	if len(data) == 0 || string(data) == "null" {
		*e = make(ExtConfig)
		return nil
	}
	return json.Unmarshal(data, e)
}

// GetString 安全读取字符串字段
func (e ExtConfig) GetString(key string) string {
	if e == nil {
		return ""
	}
	v, _ := e[key].(string)
	return v
}

// GetInt 安全读取整数字段（支持 float64 转换，JSON 数字默认为 float64）
func (e ExtConfig) GetInt(key string) int {
	if e == nil {
		return 0
	}
	switch v := e[key].(type) {
	case float64:
		return int(v)
	case int:
		return v
	case int64:
		return int(v)
	}
	return 0
}

// GetBool 安全读取布尔字段
func (e ExtConfig) GetBool(key string) bool {
	if e == nil {
		return false
	}
	v, _ := e[key].(bool)
	return v
}

// Asset 资产（服务器或中间件）
// 公共字段作为独立列，类型专属配置通过 ExtConfig（JSON）存储
type Asset struct {
	ID            uint                 `gorm:"primaryKey;autoIncrement" json:"id"`
	EnvironmentID uint                 `gorm:"not null;index" json:"environment_id"`
	GroupID       *uint                `gorm:"index" json:"group_id"`
	Category      plugin.AssetCategory `gorm:"size:20;not null;index" json:"category"`
	PluginType    string               `gorm:"size:50;not null;index" json:"plugin_type"`
	Name          string               `gorm:"size:200;not null" json:"name"`
	Description   string               `gorm:"size:500" json:"description"`
	Tags          Tags                 `gorm:"type:text" json:"tags"`
	CredentialID  *uint                `gorm:"index" json:"credential_id"`
	Status        AssetStatus          `gorm:"size:20;default:'unknown'" json:"status"`
	LastCheckedAt *time.Time           `json:"last_checked_at"`
	ExtConfig     ExtConfig            `gorm:"type:text" json:"ext_config"`
	CreatedAt     time.Time            `json:"created_at"`
	UpdatedAt     time.Time            `json:"updated_at"`

	Environment Environment `gorm:"foreignKey:EnvironmentID" json:"environment,omitempty"`
	Group       *Group      `gorm:"foreignKey:GroupID" json:"group,omitempty"`
	Credential  *Credential `gorm:"foreignKey:CredentialID" json:"credential,omitempty"`
}
